[CDAP-6103] Make default transaction timeout configurable at the (sub)program level

anew opened the pull request:
This is by passing "system.tx.timeout" as a runtime argument. (With that, we can also do it per preferences). 
- changes AbstractContext to use the default timeout from the runtime argument in its Transactional
- changes the code generation for http handlers to account for this
- changes the flowlet and custom action execution to resolve the runtime arguments for their scope (flowlet.<name>. and action.<name>.)
- changes flowlet driver to use the timeout from the runtime arguments for its transactions
- modifies the existing test case for custom tx timeouts to start with programs runtime arguments and validate that is applied.

Commit:
[CDAP-6103] Make default transaction timeout configurable at the (sub)program level

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java:
```diff
@@ -143,7 +148,24 @@ protected AbstractContext(Program program, ProgramOptions programOptions,
+    } catch (NumberFormatException e) {
+      LOG.warn("Ignoring value {} for system.data.tx.timeout because it cannot be parsed as an integer.", txTimeoutArg);
+    }
+    return txTimeout;
```
Seems like `txTimeout` can be `null`. In that case, the method needs to be annotated with `@Nullable`. Or shall we always default it to the value in cConf?

anew replied:
yes, we can do that. It means we never use the default in Tephra, but that should be ok.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java:
```diff
@@ -172,6 +194,10 @@ private MetricsContext createProgramMetrics(Program program, RunId runId,
   }
 
+  public Integer getDefaultTxTimeout() {
+    return defaultTxTimeout;
```
can this be null?

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java:
```diff
@@ -181,15 +182,34 @@ public Void call() throws Exception {
       public void execute(TxRunnable runnable) throws TransactionFailureException {
         TransactionContext txContext = datasetCache.newTransactionContext();
-        txContext.start();
+        if (defaultTimeout != null) {
```
Seems like if we have the `defaultTimeout` not being null and defaulted to `data.tx.timeout` in AbstractContext, we can save this `if-else` logic in couple places.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BasicHttpServiceContext.java:
```diff
@@ -117,4 +117,9 @@ public TransactionContext newTransactionContext() throws TransactionFailureExcep
+
+  @Override
+  public Integer getDefaultTxTimeout() {
+    return super.getDefaultTxTimeout();
```
Why override if it just call super?

anew replied:
This is tricky. The @Override is to implement the interface TransactionalHttpServiceContext, but the super implementation is from AbstractContext. I had to do that so that in here (https://github.com/caskdata/cdap/pull/7110/files#diff-7a3da93e5f2d71378a4a5e6d97dced82R83) I can get the defaultTimeout from the context. Casting to AbstractContext or BasicHttpServiceContext is not possible, because the HttpHandlerGeneratorTest uses a no-op implementation. I will add a comment.

chtyim replied:
I see. That make sense.

anew commented:
@chtyim thanks for the review. I have addressed your comments. It is not a small change because the cConf now needs to be passed to all the contexts. Hence 144 lines of change... Please take another look.

chtyim commented:
LGTM
